In [ ]:
import torch

if torch.cuda.is_available():
    device = torch.device("cuda")
    print(f"GPU model: {torch.cuda.get_device_name(0)}")
    print(f"GPU device count: {torch.cuda.device_count()}")
else:
    print("No GPU available.")
  

GPU model: Tesla T4
GPU device count: 1


# Data Pre-processing

First we will begin by pre-processing the data. In this case the sample data comes with the following features/columns: 

|   **Column**   |   **Desciption**   |
|:-:	         |:--	              |
|   STUD	|   Study Number 	|
|   DSFQ	|   Dosing Frequency	|
|   PTNM	|   Patient Number	|
|   CYCL	|    Dosing Cycles	|
|   AMT	|   Dosing Amounts	|
|   TIME	|   Time in Hours Since the Experiment Began for one Individual	|
|   TFDS	|   Time in Hours Since the Last Dosing|
|   DV/PK_timeCourse	|   The Observations of PK  	|

Let us first import the data into the notebook and observe these features of the dataset

In [2]:
from google.colab import files
 
 
uploaded = files.upload()

Saving sim_data.txt to sim_data.txt


In [4]:
# Importing required libraries for data pre-processing

import pandas as pd
import numpy as np
import io


#Reading the csv file with the data

data_complete = pd.read_csv(io.BytesIO(uploaded['sim_data.txt']))
data_complete


# data_complete = pd.read_csv("/Users/rishabhgoel/Desktop/NeuralODE_Paper_Supplementary_Code/ExampleData/sim_data.csv", na_values='.')

# data_complete

STUD   PTNM  DSFQ  CYCL    AMT    TIME    TFDS         DV
0     1000.0    1.0   3.0   1.0  296.6     0.0     0.0  20.382000
1     1000.0    1.0   3.0   1.0    0.0    24.0    24.0  73.148000
2     1000.0    1.0   3.0   1.0    0.0   216.0   216.0  19.764000
3     1000.0    1.0   3.0   1.0    0.0   504.0   504.0   3.219900
4     1000.0    1.0   3.0   2.0  288.0   504.0     0.0   3.219900
...      ...    ...   ...   ...    ...     ...     ...        ...
5371  3000.0  200.0   3.0  15.0    0.0  7416.0   360.0   6.091200
5372  3000.0  200.0   3.0  15.0    0.0  7584.0  7584.0   2.043100
5373  3000.0  200.0   3.0  17.0  259.2  8064.0     0.0   0.090115
5374  3000.0  200.0   3.0  17.0    0.0  8088.0    24.0  53.990000
5375  3000.0  200.0   3.0  17.0    0.0  8256.0   192.0  17.255000

[5376 rows x 8 columns]

There are 5376 observations and 8 features that we are looking at. Now let us begin the pre-processing the data. 

Our sample data only has the relevant columns for the model. However, in the data we receive from patients there will be many more features so it is important for us to select the correct features. Thus, let us begin by creating a place holder for the features that are important to the model.

In [5]:
#variable for colummns that we will eventually select from the raw dataset we receive

select_cols = ["STUD", "DSFQ", "PTNM", "CYCL", "AMT", "TIME", "TFDS", "DV"]

#Selecting the relevant columns from the dataframe

data_complete = data_complete[select_cols]

We then start filtering data based on multiple parameters:
    
1. Dosing Cycle < 100

In [6]:
# filtering for only the rows in the dataframe with a dosing cycle less than 100

data_complete = data_complete[data_complete.CYCL < 100]
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV
0     1000.0   3.0    1.0   1.0  296.6     0.0     0.0  20.382000
1     1000.0   3.0    1.0   1.0    0.0    24.0    24.0  73.148000
2     1000.0   3.0    1.0   1.0    0.0   216.0   216.0  19.764000
3     1000.0   3.0    1.0   1.0    0.0   504.0   504.0   3.219900
4     1000.0   3.0    1.0   2.0  288.0   504.0     0.0   3.219900
...      ...   ...    ...   ...    ...     ...     ...        ...
5371  3000.0   3.0  200.0  15.0    0.0  7416.0   360.0   6.091200
5372  3000.0   3.0  200.0  15.0    0.0  7584.0  7584.0   2.043100
5373  3000.0   3.0  200.0  17.0  259.2  8064.0     0.0   0.090115
5374  3000.0   3.0  200.0  17.0    0.0  8088.0    24.0  53.990000
5375  3000.0   3.0  200.0  17.0    0.0  8256.0   192.0  17.255000

[5376 rows x 8 columns]

We then convert the "PTNM" column to an integer type using the astype() method, and then using the map() method to apply a format string to each value in the column. The format string "{:05d}" specifies that each value should be formatted as a zero-padded integer with a width of 5 digits.

For example, if the "PTNM" column originally contained the values [1, 10, 100, 1000], after this line of code is executed, the "PTNM" column would contain the values ['00001', '00010', '00100', '01000'].

This line of code is useful for standardizing the format of the values in the "PTNM" column, which can make it easier to perform operations on the column and to compare values within the column. It can also be useful for preparing the data for downstream model applications that require the data to be in a particular format.

In [7]:
# formatting Patient Number for Unique Identifier
data_complete["PTNM"] = data_complete["PTNM"].astype("int").map("{:05d}".format)
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0  20.382000
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0  73.148000
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0  19.764000
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0   3.219900
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0   3.219900
...      ...   ...    ...   ...    ...     ...     ...        ...
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0   6.091200
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0   2.043100
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0   0.090115
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0  53.990000
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0  17.255000

[5376 rows x 8 columns]

Now we create a new column called "ID" in the pandas DataFrame data_complete. The "ID" column is being created by concatenating two existing columns in ```data_complete```: "STUD" and "PTNM".

First, the "STUD" column is being converted to an integer data type using the ```astype()``` method with the argument "int". Then, the "STUD" column is being converted to a string data type using the ```astype()``` method with the argument "str". This ensures that the values in the "STUD" column are in string format.

Next, the values in the "STUD" and "PTNM" columns are being concatenated using the + operator, which joins the two strings together. This creates a new string for each row in the DataFrame, which is assigned to the "ID" column.

For example, if the "STUD" column contained the values ['100', '101', '102', '103'] and the "PTNM" column contained the values ['00001', '00010', '00100', '01000'], then after this line of code is executed, the "ID" column would contain the values ['10000001', '10100010', '10200100', '10301000'].

This line of code is useful for creating a unique identifier for each row in the DataFrame based on the values in the "STUD" and "PTNM" columns. This can be useful for identifying and tracking individual records, as well as for linking data across multiple datasets.

In [8]:
# Creating a unique identifier column, 'ID', and formatting it based on Patient Number and Study Number
data_complete["ID"] = data_complete["STUD"].astype("int").astype("str") + data_complete["PTNM"]
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV         ID
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0  20.382000  100000001
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0  73.148000  100000001
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0  19.764000  100000001
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0   3.219900  100000001
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0   3.219900  100000001
...      ...   ...    ...   ...    ...     ...     ...        ...        ...
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0   6.091200  300000200
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0   2.043100  300000200
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0   0.090115  300000200
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0  53.990000  300000200
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0  17.255000  300000200

[5376 rows x 9 columns]

In [9]:
# Creating a dataframe with the maximum time for each patient in each study (Essentially presenting the time of the last observed dose)
time_summary = data_complete[["ID", "TIME"]].groupby("ID").max().reset_index()

# Creating a new dataframe with only the IDs for patients with a time of the last observed dose greater than 0 (eliminating errors/outliers)
selected_ptnms = time_summary[time_summary.TIME > 0].ID

# Only selecting IDs with the last observed dose greater than 0 in our main table called data_complete
data_complete = data_complete[data_complete.ID.isin(selected_ptnms)]
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS         DV         ID
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0  20.382000  100000001
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0  73.148000  100000001
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0  19.764000  100000001
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0   3.219900  100000001
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0   3.219900  100000001
...      ...   ...    ...   ...    ...     ...     ...        ...        ...
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0   6.091200  300000200
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0   2.043100  300000200
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0   0.090115  300000200
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0  53.990000  300000200
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0  17.255000  300000200

[5376 rows x 9 columns]

In [10]:
# filling in missing values in the "AMT" column with Dosing Amounts equal to 0
data_complete["AMT"] = data_complete["AMT"].fillna(0)


#Renaming the column DV to PK_timeCourse so that the name is more self-explanatory
data_complete = data_complete.rename(columns={"DV": "PK_timeCourse"})


# Duplicating PK_timeCourse column to make changes to it separately
data_complete["PK_round1"] = data_complete["PK_timeCourse"]

# Changing the PK_round1 such that if the dosing frequency is once weekly and the time of the last 
# observed dose is greater than 168hrs the PK for round 1 is zero. Similarly, if the dosing frequency 
# is once weekly and the time of the last observed dose is greater than 504hrs the PK for round 1 is zero
data_complete.loc[(data_complete.DSFQ == 1) & (data_complete.TIME >= 168), "PK_round1"] = 0
data_complete.loc[(data_complete.DSFQ == 3) & (data_complete.TIME >= 504), "PK_round1"] = 0

# filling in missing values in the "PK_round1" column with 0
data_complete["PK_round1"] = data_complete["PK_round1"].fillna(0)

# filling in missing values in the "PK_timeCourse" column with -1
data_complete["PK_timeCourse"] = data_complete["PK_timeCourse"].fillna(-1)
data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS  PK_timeCourse  \
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0      20.382000   
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0      73.148000   
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0      19.764000   
3     1000.0   3.0  00001   1.0    0.0   504.0   504.0       3.219900   
4     1000.0   3.0  00001   2.0  288.0   504.0     0.0       3.219900   
...      ...   ...    ...   ...    ...     ...     ...            ...   
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0       6.091200   
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0       2.043100   
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0       0.090115   
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0      53.990000   
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0      17.255000   

             ID  PK_round1  
0     100000001     20.382  
1     100000001     73.148  
2     100000001     19.764  
3     100000001      0.000  
4     100000001      0.000  
...         ...        ...  
5371  300000200      0.000  
5372  300000200      0.000  
5373  300000200      0.000  
5374  300000200      0.000  
5375  300000200      0.000  

[5376 rows x 10 columns]

In [11]:
# Removing rows where the "AMT" column is 0 and the "TIME" column is also 0.
data_complete = data_complete[~((data_complete.AMT == 0) & (data_complete.TIME == 0))]

# Keeping the last row for all patients with duplicate values for time in hours 
data_complete.loc[data_complete[["PTNM", "TIME"]].duplicated(keep="last"), "AMT"] = \
    data_complete.loc[data_complete[["PTNM", "TIME"]].duplicated(keep="first"), "AMT"].values

# Keeping the first row for all observations with duplicate values for all features apart from patient and time in hours 
data_complete = data_complete[~data_complete[["PTNM", "TIME"]].duplicated(keep="first")]

data_complete

STUD  DSFQ   PTNM  CYCL    AMT    TIME    TFDS  PK_timeCourse  \
0     1000.0   3.0  00001   1.0  296.6     0.0     0.0      20.382000   
1     1000.0   3.0  00001   1.0    0.0    24.0    24.0      73.148000   
2     1000.0   3.0  00001   1.0    0.0   216.0   216.0      19.764000   
3     1000.0   3.0  00001   1.0  288.0   504.0   504.0       3.219900   
5     1000.0   3.0  00001   2.0    0.0   696.0   192.0      23.210000   
...      ...   ...    ...   ...    ...     ...     ...            ...   
5371  3000.0   3.0  00200  15.0    0.0  7416.0   360.0       6.091200   
5372  3000.0   3.0  00200  15.0    0.0  7584.0  7584.0       2.043100   
5373  3000.0   3.0  00200  17.0  259.2  8064.0     0.0       0.090115   
5374  3000.0   3.0  00200  17.0    0.0  8088.0    24.0      53.990000   
5375  3000.0   3.0  00200  17.0    0.0  8256.0   192.0      17.255000   

             ID  PK_round1  
0     100000001     20.382  
1     100000001     73.148  
2     100000001     19.764  
3     100000001      0.000  
5     100000001      0.000  
...         ...        ...  
5371  300000200      0.000  
5372  300000200      0.000  
5373  300000200      0.000  
5374  300000200      0.000  
5375  300000200      0.000  

[5137 rows x 10 columns]

# Splitting Data

Now, we want to be able to split our data into training, validation, and test groups. This is a method of ensuring that we can hypertune our model without any biases and have the most robust version of our model in the final output.

We begin by defining a function called ```data_split``` that we will use eventually to split our data in a customized way. Arguments for the function include a dataframe (```df```), a column (```on_col```), a list of columns to be saved in the output dataframes (```save_cols```), the seed of the split (```seed```), and the proportion of the data that will go into the test set (```test_size```).

In [12]:
#Importing the necessary library for splitting the data into training and test sets
from sklearn.model_selection import train_test_split

#Defining the function called data_split which takes in a dataframe, a column name (likely with patient identifiers), 
# and a default test size of 20%. 
def data_split(df, on_col, save_cols=None, seed=2020, test_size=0.2):
    
    #Setting the default save_cols to all columns of the dataset
    if not save_cols:
        save_cols = df.columns.values

    #Setting a variable called target which contains all the unique values of a certain column. 
    #This variable will be helpful to split the unique patients into train and test later in the code.
    target = df[on_col].unique()
    
    #Setting the train variable to contain the random unique patients numbers of 80% of the patients 
    #while the remaining 20% is the allocated to the test variable
    train, test = train_test_split(target, random_state=seed, test_size=test_size, shuffle=True)
    
    #Creating the train and test dataframes (train_df and test_df). The training dataframe (train_df) includes all rows of 
    #patients in the train variable (train). The test dataframe (test_df) includes all rows of remaining patients (test).
    train_df = df[df[on_col].isin(train)]
    test_df = df[df[on_col].isin(test)]

    #Returning  
    return train_df[save_cols], test_df[save_cols]

We then build the code for the ```main``` that takes in our input data uses the helper function that we created above ```data_split()``` to split our data into training, validation, and test sets.

Next, the main adds a specific portion of the test set into the training and validation sets that we created above. 

Then, the main performs an augmentation that splits each patient into three groups. The main purpose of this augmentation in my opinion is to eliminate the problems associated with translating the results from one dosing regimen to the next.


In [13]:
def datasplitter(data, fold, model):
  train, test = data_split(data, "PTNM", seed=1329+fold, test_size=0.2)
  train, validate = data_split(train, "PTNM", seed=1329+fold+model, test_size=0.2)
  train = pd.concat([train, test[test.TIME < 168]], ignore_index=True)
  validate = pd.concat([validate, test[test.TIME < 168]], ignore_index=True)

  # James' augmentation
  # Deep learning is prone to overfitting, and they applied augmentation to prevent overfitting. We applied 
  # timewise truncation to increase the number of training examples. For each training example, in addition to the 
  # original example, we also truncated the examples at 1008 hr, 1512 hr, and 2016 hr and generated and added a 
  # set of new examples to the training examples.

  augment_data = pd.DataFrame(columns=train.columns)

  for ptnm in train.PTNM.unique():
      df = train[(train.PTNM == ptnm) & (train.TIME <= 2 * 21 * 24) & (train.TIME >= 0)]
      df["PTNM"] = df["PTNM"] + str(0.1)
      augment_data = pd.concat([augment_data, df], ignore_index=True)

      df = train[(train.PTNM == ptnm) & (train.TIME <= 3 * 21 * 24) & (train.TIME >= 0)]
      df["PTNM"] = df["PTNM"] + str(0.2)
      augment_data = pd.concat([augment_data, df], ignore_index=True)

      df = train[(train.PTNM == ptnm) & (train.TIME <= 4 * 21 * 24) & (train.TIME >= 0)]
      df["PTNM"] = df["PTNM"] + str(0.3)
      augment_data = pd.concat([augment_data, df], ignore_index=True)

  train = pd.concat([train, augment_data], ignore_index=True).reset_index(drop=True)
  

  train.to_csv("train.csv", index=False) 
  validate.to_csv("validate.csv", index=False)
  test.to_csv("test.csv", index=False)
  return train, validate, test

In [ ]:
# for fold in np.arange(1,6):
#   for model in np.arange(1,6):
#      train, validate, test = datasplitter(data_complete, fold, model)
#     #  runtest()
#      print(fold, model)

In [14]:
import pandas as pd
import numpy as np

def generate_interp(ptnm):
    test = pd.read_csv("test.csv")
    subset = test[test.PTNM == ptnm]
    interp_times = pd.DataFrame({"TIME": np.linspace(subset.loc[subset.TIME >= 168, "TIME"].values[:-1], \
                             subset.loc[subset.TIME >= 168, "TIME"].values[1:], 6, endpoint=False).flatten("F")})
    subset = pd.merge(subset, interp_times, how="outer", on="TIME").sort_values(by="TIME").reset_index(drop=True)
    subset["AMT"] = subset["AMT"].fillna(0)
    subset["PK_round1"] = subset["PK_round1"].fillna(0)
    subset["PTNM"] = subset["PTNM"].fillna(method="ffill")
    subset["DSFQ"] = subset["DSFQ"].fillna(1)
    subset["CYCL"] = subset["CYCL"].fillna(method="ffill")
    subset["PK_timeCourse"] = subset["PK_timeCourse"].fillna(-10) 
    
    start_time = 0
    for idx, feature in subset.iterrows():

        if feature["AMT"] > 0:
            start_time = feature["TIME"]
            continue

        subset.loc[idx, "TFDS"] = subset.loc[idx, "TIME"] - start_time

    return subset.drop(columns=["STUD", "ID"])

#ENDDDDDDD


# BRINGING IN THE ENTIRE UTILS.PY FILE

In [15]:
!pip install torchdiffeq
import os
import logging
import torch
import torch.nn as nn
import numpy as np
from sklearn.metrics import mean_squared_error, r2_score
from torchdiffeq import odeint_adjoint as odeint


def makedirs(dirname):
    if not os.path.exists(dirname):
        os.makedirs(dirname)


def get_device(tensor):
    device = torch.device("cuda")
    if tensor.is_cuda:
        device = tensor.get_device()
    return device


def load_model(ckpt_path, encoder=None, ode_func=None, classifier=None, device="cuda"):
    if not os.path.exists(ckpt_path):
        raise Exception("Checkpoint " + ckpt_path + " does not exist.")

    checkpt = torch.load(ckpt_path)
    if encoder is not None:
        encoder_state = checkpt["encoder"]
        encoder.load_state_dict(encoder_state)
        encoder.to(device)

    if ode_func is not None:
        ode_state = checkpt["ode"]
        ode_func.load_state_dict(ode_state)
        ode_func.to(device)

    if classifier is not None:
        classifier_state = checkpt["classifier"]
        classifier.load_state_dict(classifier_state)
        classifier.to(device)


def get_logger(logpath, filepath, package_files=[], displaying=True, saving=True, debug=False):
    logger = logging.getLogger()
    if debug:
        level = logging.DEBUG
    else:
        level = logging.INFO
    logger.setLevel(level)
    if saving:
        info_file_handler = logging.FileHandler(logpath, mode="a")
        info_file_handler.setLevel(level)
        logger.addHandler(info_file_handler)
    if displaying:
        console_handler = logging.StreamHandler()
        console_handler.setLevel(level)
        logger.addHandler(console_handler)
    logger.info(filepath)
    """
    with open(filepath, "r") as f:
        logger.info(f.read())

    for f in package_files:
        logger.info(f)
        with open(f, "r") as package_f:
            logger.info(package_f.read())
    """

    return logger


def inf_generator(iterable):
    """Allows training with DataLoaders in a single infinite loop:
        for i, (x, y) in enumerate(inf_generator(train_loader)):
    """
    iterator = iterable.__iter__()
    while True:
        try:
            yield iterator.__next__()
        except StopIteration:
            iterator = iterable.__iter__()


def init_network_weights(net, std = 0.1):
    for m in net.modules():
        if isinstance(m, nn.Linear):
            nn.init.normal_(m.weight, mean=0, std=std)
            nn.init.constant_(m.bias, val=0)


def reverse(tensor):
    idx = [i for i in range(tensor.size(0)-1, -1, -1)]
    return tensor[idx]


def sample_standard_gaussian(mu, sigma):
    device = get_device(mu)
    d = torch.distributions.normal.Normal(
            torch.Tensor([0.]).to(device),
            torch.Tensor([1.]).to(device))
    r = d.sample(mu.size()).squeeze(-1)
    return r * sigma.float() + mu.float()


def compute_loss_on_train(criterion, labels, preds):
    preds = preds.permute(1, 0, 2)

    idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
    # print(idx_not_nan)
    preds = preds[idx_not_nan]
    labels = labels[idx_not_nan]

    return torch.sqrt(criterion(preds, labels)).to(device)


def compute_loss_on_test(encoder, ode_func, classifier, dataloader, n_batches, device, phase):
    ptnms = []
    Times = torch.Tensor([]).to(device=device)
    predictions = torch.Tensor([]).to(device=device)
    ground_truth = torch.Tensor([]).to(device=device)
    latent_dim = 6

    for itr in range(n_batches):
        ptnm, times, features, labels, cmax_time = dataloader.__next__()
        dosing = torch.zeros([features.size(0), features.size(1), latent_dim]).to(device)
        dosing[:, :, 0] = features[:, :, -2]
        dosing = dosing.permute(1, 0, 2).to(device)

        encoder_out = encoder(features)
        qz0_mean, qz0_var = encoder_out[:, :latent_dim], encoder_out[:, latent_dim:]
        z0 = sample_standard_gaussian(qz0_mean, qz0_var)

        solves = z0.unsqueeze(0).clone()
        try:
            for idx, (time0, time1) in enumerate(zip(times[:-1], times[1:])):
                z0 += dosing[idx]
                time_interval = torch.Tensor([time0 - time0, time1 - time0]).to(device)
                sol = odeint(ode_func, z0, time_interval, rtol=1e-4, atol=1e-4)
                z0 = sol[-1].clone()
                solves = torch.cat([solves, sol[-1:, :]], 0).to(device)
        except AssertionError:
            print(times)
            print(time0, time1, time_interval, ptnm)
            continue
    
        preds = classifier(solves, cmax_time).permute(1, 0, 2)


        if phase == "test":
            idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
            preds = preds[idx_not_nan]
            labels = labels[idx_not_nan]


            times = times[idx_not_nan.flatten()]
            ptnms += ptnm * len(times)
            Times = torch.cat((Times, times*24)).to(device)

            predictions = torch.cat((predictions, preds)).to(device)
            ground_truth = torch.cat((ground_truth, labels)).to(device)
        
        else:
            """
            time_idx = (times >= 21)
            preds = preds[:, time_idx, :]
            labels = labels[:, time_idx, :]
            """
            idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
            preds = preds[idx_not_nan]
            labels = labels[idx_not_nan]

            predictions = torch.cat((predictions, preds))
            ground_truth = torch.cat((ground_truth, labels))

    rmse_loss = mean_squared_error(
        ground_truth.cpu().numpy(), predictions.cpu().numpy(),
        squared=False
    )
    r2 = r2_score(ground_truth.cpu().numpy(), predictions.cpu().numpy())

    if phase == "test":
        return {"PTNM": ptnms,
                "TIME": Times.cpu(),  
                "labels": ground_truth.cpu().tolist(), 
                "preds": predictions.cpu().tolist(),
                "loss": rmse_loss}
    else:
        return {"labels": ground_truth.cpu().tolist(), 
                "preds": predictions.cpu().tolist(),
                "loss": rmse_loss,
                "r2": r2}


def compute_loss_on_interp(encoder, ode_func, classifier, dataloader, dataloader_o, n_batches, device, phase):
    ptnms = []
    Times = torch.Tensor([]).to(device=device)
    predictions = torch.Tensor([]).to(device=device)
    ground_truth = torch.Tensor([]).to(device=device)
    latent_dim = 6

    for itr in range(n_batches):
        ptnm, times, features, labels, cmax_time = dataloader.__next__()
        ptnm_o, times_o, features_o, labels_o, cmax_time_o = dataloader_o.__next__()
        assert ptnm == ptnm_o

        dosing = torch.zeros([features.size(0), features.size(1), latent_dim]).to(device)
        dosing[:, :, 0] = features[:, :, -2]
        dosing = dosing.permute(1, 0, 2)

        encoder_out = encoder(features_o)
        qz0_mean, qz0_var = encoder_out[:, :latent_dim], encoder_out[:, latent_dim:]
        z0 = sample_standard_gaussian(qz0_mean, qz0_var)

        solves = z0.unsqueeze(0).clone()
        try:
            for idx, (time0, time1) in enumerate(zip(times[:-1], times[1:])):
                z0 += dosing[idx]
                time_interval = torch.Tensor([time0 - time0, time1 - time0]).to(device)
                sol = odeint(ode_func, z0, time_interval, rtol=1e-4, atol=1e-4)
                z0 = sol[-1].clone()
                solves = torch.cat([solves, sol[-1:, :]], 0).to(device)
        except AssertionError:
            print(times)
            print(time0, time1, time_interval, ptnm)
            continue
    
        preds = classifier(solves, cmax_time).permute(1, 0, 2)

        idx_not_nan = ~(torch.isnan(labels).to(device) | (labels == -1))
        print(idx_not_nan)
        preds = preds[idx_not_nan]
        labels = labels[idx_not_nan]

        times = times[idx_not_nan.flatten()]
        ptnms += ptnm * len(times)
        Times = torch.cat((Times, times*24)).to(device)

        predictions = torch.cat((predictions, preds)).to(device)
        ground_truth = torch.cat((ground_truth, labels)).to(device)
        
    rmse_loss = mean_squared_error(
        ground_truth.cpu().numpy(), predictions.cpu().numpy(),
        squared=False
    )
    r2 = r2_score(ground_truth.cpu().numpy(), predictions.cpu().numpy())

    return {"PTNM": ptnms,
            "TIME": Times,  
            "labels": ground_truth.cpu().tolist(), 
            "preds": predictions.cpu().tolist(),
            "loss": rmse_loss}


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


# BRINGING IN THE ENTIRE MODEL.PY FILE

In [ ]:
import torch
import torch.nn as nn
# from args import args
# from torch.nn.modules.rnn import GRU, LSTM, RNN
# import utils


class ODEFunc(nn.Module):

    def __init__(self, input_dim, hidden_dim, device=torch.device("cuda")):
        super(ODEFunc, self).__init__()

        self.net = nn.Sequential(
            nn.Linear(input_dim, hidden_dim),
            nn.SELU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.SELU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.SELU(),
            nn.Linear(hidden_dim, input_dim)
        )

        for m in self.net.modules():
            if isinstance(m, nn.Linear):
                nn.init.normal_(m.weight, mean=0, std=0.001)
                nn.init.constant_(m.bias, val=0.5)

    def forward(self, t, x):
        # print(x)
        return self.net(x)


class Encoder(nn.Module):

    def __init__(self, input_dim, output_dim, hidden_dim, device=torch.device("cuda")):
        super(Encoder, self).__init__()

        self.output_dim = output_dim
        self.input_dim = input_dim
        self.hidden_dim = hidden_dim
        self.device = device

        self.hiddens_to_output = nn.Sequential(
            nn.Linear(self.hidden_dim, self.hidden_dim),
            nn.ReLU(),
            nn.Linear(self.hidden_dim, self.output_dim),
        )
        init_network_weights(self.hiddens_to_output, std=0.001)

        # self.rnn = nn.RNN(self.input_dim, self.hidden_dim, nonlinearity="relu").to(device)
        self.rnn = nn.GRU(self.input_dim, self.hidden_dim).to(device)

    def forward(self, data):
        data = data.permute(1, 0, 2)
        data = reverse(data)
        output_rnn, _ = self.rnn(data)
        #print(output_rnn)
        outputs = self.hiddens_to_output(output_rnn[-1])
        #print(outputs)
        
        return outputs


class Classifier(nn.Module):

    def __init__(self, latent_dim, output_dim, device=torch.device("cuda")):
        super(Classifier, self).__init__()
        self.net = nn.Sequential(
            nn.Linear(latent_dim + 20, 32),
            nn.SELU(),
            nn.Linear(32, output_dim),
        )
        
        init_network_weights(self.net, std=0.001)

    def forward(self, z, cmax_time):
        cmax_time = cmax_time.repeat(z.size(0), 1, 1)
        z = torch.cat([z, cmax_time], 2)
        return self.net(z)


# BRINGING IN THE ENTIRE EVALUATION.PY FILE

In [ ]:
import pandas as pd
from sklearn.metrics import r2_score, mean_squared_error
from scipy.stats import pearsonr


def score(df, label_col, pred_col, score_fn):
    y_true = df[label_col].values
    y_pred = df[pred_col].values
    if score_fn is mean_squared_error:
        return score_fn(y_true, y_pred, squared=False)
    elif score_fn is pearsonr:
        return score_fn(y_true, y_pred)[0]
    else:
        return score_fn(y_true, y_pred)
    

def merge_predictions(files, method="mean"):
    cols = ['PTNM','TIME','preds']
    left = pd.read_csv(files[0])
    for f in files[1:]:
        right = pd.read_csv(f)
        left = left.merge(right[cols], on=["PTNM", "TIME"], how="left")
    preds = [col for col in left.columns.values if col.startswith("preds")]
    left["pred_agg"] = left[preds].agg(method, axis=1)
    left = left[left.TIME >= 168]
    print(left.shape)
    return left

def write_score(pred_df, fold, model_name, rmse_dict, score_fn):
    res = score(pred_df, "labels", "pred_agg", score_fn)
    if not rmse_dict.get("fold", []) or rmse_dict["fold"][-1] != fold:
        rmse_dict["fold"] = rmse_dict.get("fold", []) + [fold]
    rmse_dict[model_name] = rmse_dict.get(model_name, []) + [round(res, 5)]


def main(score_type, score_fn, folds, models):
    records = {}
    for fold in folds:
        in_file = ["fold_{}/fold_{}_model_{}.csv".format(fold, fold, m) for m in models]
        predictions = merge_predictions(in_file)
        predictions.to_csv("predictions.csv", index=False)
        write_score(predictions, fold, "ensemble (mean)", records, score_fn)

        for f in in_file:
            predictions = merge_predictions([f])
            write_score(predictions, fold, f.split("/")[1], records, score_fn)
    
    df = pd.DataFrame(records)
    print(df)
    summary_df = df.drop(columns="fold").agg(["min", "max", "mean", "median"])
    print(summary_df)
    df.to_csv(f"{score_type}.txt", sep="\t", index=False)
    summary_df.to_csv(f"{score_type}_summary.txt", sep="\t", index=False)


  # main("rmse", mean_squared_error, args)
  # main("r2", r2_score, args)
  # main("correlation", pearsonr, args)


# BRINGING IN THE ENTIRE DATA_PARSE.PY FILE

In [ ]:
import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np
import pandas as pd
import random
from sklearn.model_selection import train_test_split

class TDM1(Dataset):

    def __init__(self, data_to_load, label_col, feature_cols, device, phase="train"):
        self.data = pd.read_csv(data_to_load)

        self.label_col = label_col
        self.features = feature_cols
        self.device = device
        self.phase = phase
        self.data["TIME"] = self.data["TIME"] / 24

    def __len__(self):
        return self.data.PTNM.unique().shape[0] 

    def __getitem__(self, index):
        ptnm = self.data.PTNM.unique()[index]
        cur_data = self.data[self.data["PTNM"] == ptnm]
        times, features, labels, cmax_time = self.process(cur_data)
        return ptnm, times, features, labels, cmax_time

    def process(self, data):
        data = data.reset_index(drop=True)
        """
        if self.phase == "train":
            random_time = np.random.randint(low=21, high=int(data["TIME"].max()) + 1)
            data = data[data.TIME <= random_time]
        else:
            pass
        """
        if (data.DSFQ == 1).all():
            cmax_time = data.loc[data.TIME < 7, ["TIME", "PK_timeCourse"]].values.flatten()
            cmax = data.loc[data.TIME < 7, "PK_timeCourse"].max()
        else:
            cmax_time = data.loc[data.TIME < 21, ["TIME", "PK_timeCourse"]].values.flatten()
            cmax = data.loc[data.TIME < 21, "PK_timeCourse"].max()

        cmax_time_full = np.zeros((20, ))
        if len(cmax_time) <= 20:
            cmax_time_full[:len(cmax_time)] = cmax_time
        else:
            cmax_time_full[:] = cmax_time[:20]
        
        data["PK_round1"] = data["PK_round1"] / cmax

        features = data[self.features].values
        labels = data[self.label_col].values
        times = data["TIME"].values

        times = torch.from_numpy(times)
        features = torch.from_numpy(features)
        labels = torch.from_numpy(labels).unsqueeze_(-1)
        cmax_time_full = torch.from_numpy(cmax_time_full)
        return times, features, labels, cmax_time_full


def tdm1_collate_fn(batch, device=device):
    D = batch[0][2].shape[1]
    N = 1

    combined_tt, inverse_indices = torch.unique(torch.cat([ex[1] for ex in batch]),
        sorted=True, return_inverse=True)
    combined_tt = combined_tt.to(device)
    # print(combined_tt, inverse_indices)

    offset = 0
    combined_features = torch.zeros([len(batch), len(combined_tt), D]).to(device)
    combined_label = torch.zeros([len(batch), len(combined_tt), N]).to(device)
    combined_label[:] = np.nan
    combined_cmax_time = torch.zeros([len(batch), 20]).to(device)
    # print(combined_label.shape)

    ptnms = []
    for b, (ptnm, tt, features, label, cmax_time) in enumerate(batch):
        ptnms.append(ptnm)
        tt = tt.to(device)
        features = features.to(device)
        label = label.to(device)
        cmax_time = cmax_time.to(device)

        indices = inverse_indices[offset:offset + len(tt)]
        offset += len(tt)

        combined_features[b, indices] = features.float()
        combined_label[b, indices] = label.float()
        combined_cmax_time[b, :] = cmax_time.float()
    combined_tt = combined_tt.float()

    return ptnms, combined_tt, combined_features, combined_label, combined_cmax_time


def parse_tdm1(device, phase="train"):
    train_data_path = "train.csv"
    val_data_path = "validate.csv"
    test_data_path = "test.csv"

    feature_cols = ['TFDS','TIME','CYCL','AMT',"PK_round1"]
    """
    covariates = ['SEX','AGE','WT','RACR','RACE','BSA',
                  'BMI','ALBU','TPRO','WBC','CRCL','CRET',
                  'SGOT','SGPT','TBIL','TMBD','ALKP', 'HER', 
                  'ECOG','KEOALL','ASIAN']
    feature_cols += covariates
    """
    label_col = "PK_timeCourse"
    train = TDM1(train_data_path, label_col, feature_cols, device, phase="train")
    validate = TDM1(val_data_path, label_col, feature_cols, device, phase="validate")
    test = TDM1(test_data_path, label_col, feature_cols, device, phase="test")

    ptnm, times, features, labels, cmax_time = train[0]
    input_dim = features.size(-1)
    # n_labels = 1

    if phase == "train":
        train_dataloader = DataLoader(train, batch_size=1, shuffle=True, 
            collate_fn=lambda batch: tdm1_collate_fn(batch, device))
        val_dataloader = DataLoader(validate, batch_size=1, shuffle=False,
            collate_fn=lambda batch: tdm1_collate_fn(batch, device))

        dataset_objs = {
            "train_dataloader": inf_generator(train_dataloader),
            "val_dataloader": inf_generator(val_dataloader),
            "n_train_batches": len(train_dataloader),
            "n_val_batches": len(val_dataloader),
            "input_dim": input_dim
        }

    else:
        test_dataloader = DataLoader(test, batch_size=1, shuffle=False,
            collate_fn=lambda batch: tdm1_collate_fn(batch, device))

        dataset_objs = {
            "test_dataloader": inf_generator(test_dataloader),
            "n_test_batches": len(test_dataloader),
            "input_dim": input_dim
        }

    return dataset_objs

    




# if __name__ == "__main__":
#     print("run")
#     data = parse_tdm1("cpu")
#     for ptnms, times, features, labels, cmax_time in data["train_dataloader"]:
#         print(ptnms)
#         print(times)
#         print(features)
#         print(labels)
#         print(cmax_time)
#         break


# BRINGING IN THE ENTIRE RUNTRAIN.PY FILE

In [ ]:
import os
import sys
import numpy as np
import pandas as pd
from random import SystemRandom
from tqdm import tqdm

import torch
import torch.nn as nn
import torch.optim as optim
from torchdiffeq import odeint_adjoint as odeint
from scipy.stats import pearsonr

# import utils
# from model import *
# from data_parse import parse_tdm1
# from args import args


device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
file_name = os.path.basename("__file__")[:-3]


input_cmd = sys.argv
input_cmd = " ".join(input_cmd)

########################################################################
## Main runnings
def runtraining(random_seed, model, fold, lr=0.00005, l2=0.1, epochs=30, tol=1e-4, continue_train=False):
  save = f"Fold{fold}"
  makedirs(save)
  torch.manual_seed(random_seed + model + fold)
  np.random.seed(random_seed + model + fold)

  ckpt_path = os.path.join(save, f"fold_{fold}_model_{model}.ckpt")
  ########################################################################
  tdm1_obj = parse_tdm1(device, phase="train")
  input_dim = tdm1_obj["input_dim"]
  hidden_dim = 128
  latent_dim = 6

  encoder = Encoder(input_dim=input_dim, output_dim=2 * latent_dim, hidden_dim=hidden_dim).to(device)
  ode_func = ODEFunc(input_dim=latent_dim, hidden_dim=16).to(device)
  classifier = Classifier(latent_dim=latent_dim, output_dim=1).to(device)

  # torch.save({'encoder': encoder.state_dict(), 'ode': ode_func.state_dict(), 'classifier': classifier.state_dict()}, ckpt_path)

  if continue_train:
      load_model(ckpt_path, encoder, ode_func, classifier, device)

  ########################################################################
  ## Train
  log_path = "logs/" + f"fold_{fold}_model_{model}.log"
  makedirs("logs/")
  logger = get_logger(logpath=log_path, filepath=os.path.abspath("__file__"))
  logger.info(input_cmd)

  batches_per_epoch = tdm1_obj["n_train_batches"]
  criterion = nn.MSELoss().to(device=device)
  params = (list(encoder.parameters()) + 
            list(ode_func.parameters()) + 
            list(classifier.parameters()))
  optimizer = optim.Adam(params, lr=lr, weight_decay=l2)
  best_rmse = 0x7fffffff
  best_epochs = 0

  for epoch in range(1, epochs):

      for _ in tqdm(range(batches_per_epoch), ascii=True):
          optimizer.zero_grad()

          ptnms, times, features, labels, cmax_time = tdm1_obj["train_dataloader"].__next__()
          dosing = torch.zeros([features.size(0), features.size(1), latent_dim]).to(device)
          dosing[:, :, 0] = features[:, :, -2]
          dosing = dosing.permute(1, 0, 2)

          encoder_out = encoder(features)
          qz0_mean, qz0_var = encoder_out[:, :latent_dim], encoder_out[:, latent_dim:]
          z0 = sample_standard_gaussian(qz0_mean, qz0_var).to(device)
          
          solves = z0.unsqueeze(0).clone()
          try:
              for idx, (time0, time1) in enumerate(zip(times[:-1], times[1:])):
                  z0 += dosing[idx]
                  # print(z0.device)
                  time_interval = torch.Tensor([time0 - time0, time1 - time0]).to(device)
                  sol = odeint(ode_func, z0, time_interval, rtol=tol, atol=tol).to(device)
                  z0 = sol[-1].clone()
                  solves = torch.cat([solves, sol[-1:, :]], 0).to(device)
          except AssertionError:
              print(times)
              print(time0, time1, time_interval, ptnms)
              continue
      
          preds = classifier(solves, cmax_time)

          loss = compute_loss_on_train(criterion, labels, preds)
          try: 
              loss.backward()
          except RuntimeError:
              print(ptnms)
              print(times)
              continue
          optimizer.step()
      
      idx_not_nan = ~(torch.isnan(labels) | (labels == -1))
      preds = preds.permute(1, 0, 2)[idx_not_nan]
      labels = labels[idx_not_nan]
      print(preds)
      print(labels)

      with torch.no_grad():
          
          train_res = compute_loss_on_test(encoder, ode_func, classifier,
              tdm1_obj["train_dataloader"], tdm1_obj["n_train_batches"], 
              device, phase="train")

          validation_res = compute_loss_on_test(encoder, ode_func, classifier,
              tdm1_obj["val_dataloader"], tdm1_obj["n_val_batches"], 
              device, phase="validate")
          
          train_loss = train_res["loss"] 
          validation_loss = validation_res["loss"]
          if validation_loss < best_rmse:
              torch.save({'encoder': encoder.state_dict(),
                          'ode': ode_func.state_dict(),
                          'classifier': classifier.state_dict()}, ckpt_path)
              best_rmse = validation_loss
              best_epochs = epoch

          message = """
          Epoch {:04d} | Training loss {:.6f} | Training R2 {:.6f} | Validation loss {:.6f} | Validation R2 {:.6f}
          Best loss {:.6f} | Best epoch {:04d}
          """.format(epoch, train_loss, train_res["r2"], validation_loss, validation_res["r2"], best_rmse, best_epochs)
          logger.info(message)
            

#BRINGING IN THE ENTIRE RUNPREDICT.PY FILE

In [ ]:
import os
import numpy as np
import pandas as pd
import torch



device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

###############################################################
## Main runnings
def runpredict(random_seed, model, fold, lr=0.00005, l2=0.1, epochs=30, tol=1e-4, continue_train=True):
  save = f"Fold{fold}"
  makedirs(save)
  ckpt_path = os.path.join(save, f"fold_{fold}_model_{model}.ckpt")
  eval_path = os.path.join(save, f"fold_{fold}_model_{model}.csv")
  res_path = "rmse.csv"

  ########################################################################
  tdm1_obj = parse_tdm1(device, phase="test")
  input_dim = tdm1_obj["input_dim"]
  hidden_dim = 128 
  latent_dim = 6

  encoder = Encoder(input_dim=input_dim, output_dim=2 * latent_dim, hidden_dim=hidden_dim)
  ode_func = ODEFunc(input_dim=latent_dim, hidden_dim=16)
  classifier = Classifier(latent_dim=latent_dim, output_dim=1)

  load_model(ckpt_path, encoder, ode_func, classifier, device)

  ########################################################################
  ## Predict & Evaluate
  with torch.no_grad():
      test_res = compute_loss_on_test(encoder, ode_func, classifier,
          tdm1_obj["test_dataloader"], tdm1_obj["n_test_batches"], 
          device, phase="test")
      

  eval_results = pd.DataFrame(test_res).drop(columns="loss")
  eval_results.to_csv(eval_path, index=False)

  """
  with torch.no_grad():
      test_res = utils.compute_loss_on_interp(encoder, ode_func, classifier, args,
          tdm1_obj["interp_dataloader"], tdm1_obj["test_dataloader"], tdm1_obj["n_interp_batches"], 
          device, phase="test")

  eval_results = pd.DataFrame(test_res).drop(columns="loss")
  eval_results.to_csv(eval_path + ".interp", index=False)

  with torch.no_grad():
      test_res = utils.compute_loss_on_interp(encoder, ode_func, classifier, args,
          tdm1_obj["nodosing_dataloader"], tdm1_obj["test_dataloader"], tdm1_obj["n_interp_batches"], 
          device, phase="test")

  eval_results = pd.DataFrame(test_res).drop(columns="loss")
  eval_results.to_csv(eval_path + ".nodosing", index=False)
  """

RUNNING THE NEURAL-ODE MODEL

In [ ]:
fold, model = 1,1
train, validate, test = datasplitter(data_complete, fold, model)
ptnms = ['5010','5011','5012','5013','5014','5015','5016','5209','5210','5211','5212','5409','5410','5411','5412','5413','5414','5415','5416','5602','5603','5604','5605','5606','5607','5608','5609','5610']
test_interp = pd.DataFrame()
for ptnm in ptnms:
    print(ptnm)
    test_interp = pd.concat([test_interp, generate_interp(int(ptnm))], ignore_index=True)
test_interp.to_csv("test_interp.csv", index=False)
test = pd.read_csv("test.csv")
test.loc[test.CYCL > 5, "AMT"] = 0
test.to_csv("test_nodosing.csv", index=False)
runtraining(1000, model, fold, 0.00005, 0.1, 30, 1e-4, False)
runpredict(1000, model, fold, 0.00005, 0.1, 30, 1e-4, True)

print(fold, model)

<ipython-input-13-a1547cc7bd9e>:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.1)
<ipython-input-13-a1547cc7bd9e>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.2)
<ipython-input-13-a1547cc7bd9e>:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide

5010
5011
5012
5013
5014
5015
5016
5209
5210
5211
5212
5409
5410
5411
5412
5413
5414
5415
5416
5602
5603
5604
5605
5606
5607
5608
5609
5610


INFO:root:/content/__file__
/content/__file__
INFO:root:/usr/local/lib/python3.8/dist-packages/ipykernel_launcher.py -f /root/.local/share/jupyter/runtime/kernel-395d1998-8e4e-4dc9-a7fa-cdd8edebe74a.json
/usr/local/lib/python3.8/dist-packages/ipykernel_launcher.py -f /root/.local/share/jupyter/runtime/kernel-395d1998-8e4e-4dc9-a7fa-cdd8edebe74a.json
100%|##########| 672/672 [04:43<00:00,  2.37it/s]


tensor([ 0.2391,  5.1717,  8.8703,  9.2620, 12.9289, 13.1532, 13.3494, 17.0739,
        17.2425], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 5.8156,  0.8816, 16.1310,  0.8871, 19.8420,  3.7824,  0.8870, 20.1590,
         5.8156], device='cuda:0')


INFO:root:
          Epoch 0001 | Training loss 26.341450 | Training R2 -0.768458 | Validation loss 28.023655 | Validation R2 -1.255971
          Best loss 28.023655 | Best epoch 0001
          

          Epoch 0001 | Training loss 26.341450 | Training R2 -0.768458 | Validation loss 28.023655 | Validation R2 -1.255971
          Best loss 28.023655 | Best epoch 0001
          
100%|##########| 672/672 [04:40<00:00,  2.39it/s]


tensor([0.9732, 4.5797, 4.7871, 7.2709, 7.4598], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([0.1803, 6.8826, 0.1879, 5.1060, 0.1803], device='cuda:0')


INFO:root:
          Epoch 0002 | Training loss 27.136173 | Training R2 -0.876777 | Validation loss 29.782393 | Validation R2 -1.548022
          Best loss 28.023655 | Best epoch 0001
          

          Epoch 0002 | Training loss 27.136173 | Training R2 -0.876777 | Validation loss 29.782393 | Validation R2 -1.548022
          Best loss 28.023655 | Best epoch 0001
          
100%|##########| 672/672 [04:38<00:00,  2.41it/s]


tensor([1.3512, 2.2844, 2.2852, 2.2861, 2.2869, 2.2901, 2.2925, 2.2958, 2.2982,
        2.3006, 3.0193, 3.0262, 3.0282, 3.0332, 3.0391, 3.6152, 3.6231, 3.6309,
        3.6376], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([0.1027, 4.3675, 3.3207, 2.6789, 2.2198, 1.1160, 0.6726, 0.3426, 0.2066,
        0.1245, 4.4427, 1.1442, 0.8164, 0.3513, 0.1277, 4.4458, 1.1451, 0.3516,
        0.1278], device='cuda:0')


INFO:root:
          Epoch 0003 | Training loss 26.767849 | Training R2 -0.826175 | Validation loss 29.269934 | Validation R2 -1.461090
          Best loss 28.023655 | Best epoch 0001
          

          Epoch 0003 | Training loss 26.767849 | Training R2 -0.826175 | Validation loss 29.269934 | Validation R2 -1.461090
          Best loss 28.023655 | Best epoch 0001
          
100%|##########| 672/672 [04:37<00:00,  2.42it/s]


tensor([ 4.1092,  9.2334,  9.0717,  8.9328, 13.0813, 12.9160, 12.7743],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 1.3031, 17.1530,  4.2189,  1.2679, 17.6280,  4.3359,  1.3031],
       device='cuda:0')


INFO:root:
          Epoch 0004 | Training loss 24.577896 | Training R2 -0.539589 | Validation loss 25.545418 | Validation R2 -0.874606
          Best loss 25.545418 | Best epoch 0004
          

          Epoch 0004 | Training loss 24.577896 | Training R2 -0.539589 | Validation loss 25.545418 | Validation R2 -0.874606
          Best loss 25.545418 | Best epoch 0004
          
100%|##########| 672/672 [04:39<00:00,  2.41it/s]


tensor([13.4741, 18.5694, 17.9659, 16.8139, 21.7352, 20.9481, 19.4488],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([17.6690, 69.9310, 19.6240,  2.0229, 70.4890, 19.8330,  2.0444],
       device='cuda:0')


INFO:root:
          Epoch 0005 | Training loss 18.737062 | Training R2 0.105216 | Validation loss 17.964836 | Validation R2 0.072892
          Best loss 17.964836 | Best epoch 0005
          

          Epoch 0005 | Training loss 18.737062 | Training R2 0.105216 | Validation loss 17.964836 | Validation R2 0.072892
          Best loss 17.964836 | Best epoch 0005
          
100%|##########| 672/672 [04:37<00:00,  2.42it/s]


tensor([10.7882, 18.6375], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 7.8802, 54.9540], device='cuda:0')


INFO:root:
          Epoch 0006 | Training loss 18.679979 | Training R2 0.110659 | Validation loss 17.838053 | Validation R2 0.085931
          Best loss 17.838053 | Best epoch 0006
          

          Epoch 0006 | Training loss 18.679979 | Training R2 0.110659 | Validation loss 17.838053 | Validation R2 0.085931
          Best loss 17.838053 | Best epoch 0006
          
100%|##########| 672/672 [04:40<00:00,  2.40it/s]


tensor([ 3.4977, 10.9911,  9.3939,  7.9964, 13.2339,  7.3523, 12.4247,  5.2493,
        10.0115,  3.1056], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 2.5158, 10.0050,  2.5792,  0.8070, 40.4740,  0.8047, 40.4720,  0.8047,
        39.6060,  0.7875], device='cuda:0')


INFO:root:
          Epoch 0007 | Training loss 18.294975 | Training R2 0.146941 | Validation loss 17.784164 | Validation R2 0.091446
          Best loss 17.784164 | Best epoch 0007
          

          Epoch 0007 | Training loss 18.294975 | Training R2 0.146941 | Validation loss 17.784164 | Validation R2 0.091446
          Best loss 17.784164 | Best epoch 0007
          
100%|##########| 672/672 [04:45<00:00,  2.35it/s]


tensor([ 5.0489,  9.5644,  9.1430, 15.3729,  8.7122,  7.6244, 13.7716,  5.7963,
         5.1712,  4.2042,  3.8961,  5.3573,  3.0964,  4.1691,  3.0957],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 8.4665,  2.6608,  2.2691, 18.5760,  3.2225,  2.3435, 25.5770,  3.2261,
         2.3462,  3.2262,  2.3463, 25.6730,  2.3549, 30.2890,  8.4665],
       device='cuda:0')


INFO:root:
          Epoch 0008 | Training loss 18.257090 | Training R2 0.150470 | Validation loss 17.998465 | Validation R2 0.069417
          Best loss 17.784164 | Best epoch 0007
          

          Epoch 0008 | Training loss 18.257090 | Training R2 0.150470 | Validation loss 17.998465 | Validation R2 0.069417
          Best loss 17.784164 | Best epoch 0007
          
100%|##########| 672/672 [04:42<00:00,  2.38it/s]


tensor([19.9414, 45.5474], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([19.2920, 85.6310], device='cuda:0')


INFO:root:
          Epoch 0009 | Training loss 16.834253 | Training R2 0.277724 | Validation loss 17.300528 | Validation R2 0.140190
          Best loss 17.300528 | Best epoch 0009
          

          Epoch 0009 | Training loss 16.834253 | Training R2 0.277724 | Validation loss 17.300528 | Validation R2 0.140190
          Best loss 17.300528 | Best epoch 0009
          
100%|##########| 672/672 [04:41<00:00,  2.39it/s]


tensor([ 8.6609, 27.2044, 22.9676, 18.6471, 14.8763,  8.7254, 13.0352,  8.5347,
         6.6915,  8.4333,  7.3392,  6.6012,  6.5200,  7.8282,  7.1562,  6.4779,
         6.3803,  7.5840,  6.7952,  5.9959,  5.8809,  7.0077,  6.0941,  5.1688,
         5.0257,  6.0533,  5.2046,  4.7167,  5.3301,  4.8579,  4.3112,  4.8474,
         4.3210,  3.7116,  4.2703,  3.7692], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 7.2829, 42.8440, 12.3400,  4.1313,  1.6174,  1.6724, 20.9970,  5.9940,
         1.7166, 17.7690,  5.9424,  1.9898,  1.7019, 17.6790,  5.9126,  1.9798,
         1.6933, 17.6670,  5.9084,  1.9784,  1.6921, 17.9310,  5.9966,  2.0079,
         1.7174, 17.9400,  5.9999,  1.7183, 17.9410,  6.0000,  1.7184, 17.4120,
         5.8233,  1.6677, 18.6260,  7.2829], device='cuda:0')


INFO:root:
          Epoch 0010 | Training loss 16.071266 | Training R2 0.341713 | Validation loss 16.373804 | Validation R2 0.229836
          Best loss 16.373804 | Best epoch 0010
          

          Epoch 0010 | Training loss 16.071266 | Training R2 0.341713 | Validation loss 16.373804 | Validation R2 0.229836
          Best loss 16.373804 | Best epoch 0010
          
100%|##########| 672/672 [04:42<00:00,  2.37it/s]


tensor([17.9965, 27.2236, 26.5087, 25.7919], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([31.1600, 21.5030, 17.4770, 14.5830], device='cuda:0')


INFO:root:
          Epoch 0011 | Training loss 15.341714 | Training R2 0.400122 | Validation loss 15.422334 | Validation R2 0.316743
          Best loss 15.422334 | Best epoch 0011
          

          Epoch 0011 | Training loss 15.341714 | Training R2 0.400122 | Validation loss 15.422334 | Validation R2 0.316743
          Best loss 15.422334 | Best epoch 0011
          
100%|##########| 672/672 [04:41<00:00,  2.39it/s]


tensor([14.3396, 44.7900, 43.5865, 42.3810, 41.1735, 37.5388, 56.6964, 52.9522,
        45.4368, 63.3831, 55.7440, 48.0819, 45.2357, 57.9942, 41.0654, 50.5978,
        37.1669, 42.2982, 34.6367, 39.4978, 33.3567, 37.6751, 31.0308, 34.0545,
        31.2226, 33.9361, 31.2001, 33.7443, 31.7898], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([20.7140, 46.6930, 36.3750, 29.6610, 24.7630, 15.1220, 59.5840, 39.0430,
        20.0970, 63.5550, 35.3180, 21.6870, 21.4910, 52.0110, 22.1980, 50.6250,
        21.6590, 41.1430, 21.2610, 42.8590, 22.1220, 43.3930, 22.4050, 41.0620,
        21.2290, 40.3330, 20.8420, 40.0930, 20.7140], device='cuda:0')


INFO:root:
          Epoch 0012 | Training loss 14.818567 | Training R2 0.440335 | Validation loss 14.742236 | Validation R2 0.375675
          Best loss 14.742236 | Best epoch 0012
          

          Epoch 0012 | Training loss 14.818567 | Training R2 0.440335 | Validation loss 14.742236 | Validation R2 0.375675
          Best loss 14.742236 | Best epoch 0012
          
100%|##########| 672/672 [04:42<00:00,  2.38it/s]


tensor([ 9.0743, 30.1433, 10.3643, 30.3810, 17.6184,  9.6584, 17.0900, 14.0828,
        10.7773, 14.1977, 11.7057, 14.7096, 11.8826, 11.6649, 14.4221, 11.7998,
        13.6988, 12.6351, 12.5533, 14.3729, 13.0718, 14.8133, 13.4604, 13.3564,
        13.3852, 14.9641, 13.1994, 12.7226, 14.1258, 13.0930, 12.0605, 13.4694,
        12.3585], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 9.4905, 23.3070,  3.3688, 62.3120, 21.5050,  3.5695, 43.4420, 24.6750,
         3.5667, 25.0220,  3.6169, 24.3010,  4.0329,  3.5126, 24.4200,  3.5299,
        24.5050,  4.0669,  3.5422, 24.5080,  3.5425, 24.5080,  4.0673,  3.5426,
         3.5090, 24.4200,  3.5298,  3.4397, 24.1540,  9.1820,  3.4914, 24.9650,
         9.4905], device='cuda:0')


INFO:root:
          Epoch 0013 | Training loss 14.444537 | Training R2 0.468231 | Validation loss 14.422903 | Validation R2 0.402429
          Best loss 14.422903 | Best epoch 0013
          

          Epoch 0013 | Training loss 14.444537 | Training R2 0.468231 | Validation loss 14.422903 | Validation R2 0.402429
          Best loss 14.422903 | Best epoch 0013
          
100%|##########| 672/672 [04:41<00:00,  2.39it/s]


tensor([ 2.9864, 23.8703,  7.7314, 22.6760, 14.1833,  8.5523, 15.4518, 12.8690,
        10.6400,  9.7570], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 0.9887, 13.2440,  1.0495, 44.2920, 11.1530,  1.0742, 43.9160, 13.4380,
         2.8234,  1.0649], device='cuda:0')


INFO:root:
          Epoch 0014 | Training loss 14.236508 | Training R2 0.483438 | Validation loss 14.391095 | Validation R2 0.405062
          Best loss 14.391095 | Best epoch 0014
          

          Epoch 0014 | Training loss 14.236508 | Training R2 0.483438 | Validation loss 14.391095 | Validation R2 0.405062
          Best loss 14.391095 | Best epoch 0014
          
100%|##########| 672/672 [04:42<00:00,  2.38it/s]


tensor([12.9101, 55.2716, 25.1442, 11.1457, 22.6034], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([16.1690, 59.8340,  5.3050,  1.4892, 16.1690], device='cuda:0')


INFO:root:
          Epoch 0015 | Training loss 13.887590 | Training R2 0.508448 | Validation loss 13.889879 | Validation R2 0.445782
          Best loss 13.889879 | Best epoch 0015
          

          Epoch 0015 | Training loss 13.887590 | Training R2 0.508448 | Validation loss 13.889879 | Validation R2 0.445782
          Best loss 13.889879 | Best epoch 0015
          
100%|##########| 672/672 [04:46<00:00,  2.34it/s]


tensor([33.4062, 45.7921, 44.3150, 42.8444, 38.4714, 51.1083, 44.6173, 38.2282,
        41.4508, 33.1183, 43.8642, 26.2220, 34.4036, 24.6529, 30.0550, 23.5997,
        28.8094, 21.6657, 26.8286, 18.9788, 24.0048, 18.7110, 20.9831],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([66.9750, 34.8300, 30.4850, 27.1220, 19.3810, 59.4210, 39.4690, 28.2220,
        45.4640, 32.5130, 71.1690, 34.2430, 72.7160, 35.0360, 72.9860, 35.1970,
        73.1260, 35.2700, 73.7120, 35.5420, 73.9550, 35.6660, 66.9750],
       device='cuda:0')


INFO:root:
          Epoch 0016 | Training loss 13.523159 | Training R2 0.533908 | Validation loss 13.310001 | Validation R2 0.491091
          Best loss 13.310001 | Best epoch 0016
          

          Epoch 0016 | Training loss 13.523159 | Training R2 0.533908 | Validation loss 13.310001 | Validation R2 0.491091
          Best loss 13.310001 | Best epoch 0016
          
100%|##########| 672/672 [04:44<00:00,  2.36it/s]


tensor([13.6608, 73.8658, 50.0054, 12.0001, 60.0117, 30.5455, 13.8400, 39.0830,
        27.2152, 13.7561], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([18.0000, 80.3400, 25.8200,  4.2027, 84.0030, 27.2700,  4.4388, 82.8660,
        26.9150,  4.3809], device='cuda:0')


INFO:root:
          Epoch 0017 | Training loss 13.309327 | Training R2 0.548531 | Validation loss 13.026054 | Validation R2 0.512573
          Best loss 13.026054 | Best epoch 0017
          

          Epoch 0017 | Training loss 13.309327 | Training R2 0.548531 | Validation loss 13.026054 | Validation R2 0.512573
          Best loss 13.026054 | Best epoch 0017
          
100%|##########| 672/672 [04:45<00:00,  2.35it/s]


tensor([ 8.2451, 25.4013, 11.7649,  8.2381, 27.4235, 18.4464, 15.8813, 12.6014,
         8.5868], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([17.2550, 16.6050,  5.5696,  2.1836, 53.8670, 17.2450, 10.7980,  5.7840,
         2.2677], device='cuda:0')


INFO:root:
          Epoch 0018 | Training loss 13.005368 | Training R2 0.568917 | Validation loss 12.941817 | Validation R2 0.518857
          Best loss 12.941817 | Best epoch 0018
          

          Epoch 0018 | Training loss 13.005368 | Training R2 0.568917 | Validation loss 12.941817 | Validation R2 0.518857
          Best loss 12.941817 | Best epoch 0018
          
100%|##########| 672/672 [04:46<00:00,  2.35it/s]


tensor([-3.2930, 41.1100, 12.9715,  9.2845,  7.5195, 30.1203, 22.5275, 13.2624,
        11.1938, 29.7901, 17.8630, 15.5830, 12.9240, 21.6705, 17.9947, 14.4187],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 0.7684, 22.7900,  2.1049,  0.8118,  0.5042, 22.8300,  5.4673,  0.8133,
         0.5051, 28.9870,  4.3086,  1.6618,  0.5051, 21.5700,  5.1656,  0.7684],
       device='cuda:0')


INFO:root:
          Epoch 0019 | Training loss 12.462146 | Training R2 0.604177 | Validation loss 12.138934 | Validation R2 0.576703
          Best loss 12.138934 | Best epoch 0019
          

          Epoch 0019 | Training loss 12.462146 | Training R2 0.604177 | Validation loss 12.138934 | Validation R2 0.576703
          Best loss 12.138934 | Best epoch 0019
          
100%|##########| 672/672 [04:48<00:00,  2.33it/s]


tensor([ 3.2754, 64.2589, 37.9990,  3.2725, 24.9393,  5.3459], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 1.5969, 68.9360, 20.1160,  1.3031, 20.2650,  1.3127], device='cuda:0')


INFO:root:
          Epoch 0020 | Training loss 12.032800 | Training R2 0.630981 | Validation loss 11.731411 | Validation R2 0.604647
          Best loss 11.731411 | Best epoch 0020
          

          Epoch 0020 | Training loss 12.032800 | Training R2 0.630981 | Validation loss 11.731411 | Validation R2 0.604647
          Best loss 11.731411 | Best epoch 0020
          
100%|##########| 672/672 [04:50<00:00,  2.31it/s]


tensor([ 2.9406, 61.2473, 28.9992,  3.8048, 22.1847,  5.6358], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 2.0389, 55.6560, 16.1950,  2.1895, 15.9180,  2.1521], device='cuda:0')


INFO:root:
          Epoch 0021 | Training loss 11.403233 | Training R2 0.668586 | Validation loss 11.126586 | Validation R2 0.644362
          Best loss 11.126586 | Best epoch 0021
          

          Epoch 0021 | Training loss 11.403233 | Training R2 0.668586 | Validation loss 11.126586 | Validation R2 0.644362
          Best loss 11.126586 | Best epoch 0021
          
100%|##########| 672/672 [04:51<00:00,  2.31it/s]


tensor([ 1.8661, 33.6066, 13.1035,  3.6057, -0.9194, -1.3781, 10.9742,  3.5218,
        -0.2657, -0.7849, 10.3535,  3.6867, -0.8698, -1.3888,  9.7753,  2.3312,
        -1.1604, -1.5095,  9.1411,  2.5107, -0.7796,  6.1673,  3.0142, -0.1440,
         7.4770,  3.0706,  0.8500,  0.5366,  2.6362,  0.9857, 11.5149,  6.6135,
         3.2692,  1.5496], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 0.4168, 38.0910,  7.1683,  1.7787,  0.4414,  0.3499, 11.5080,  1.7944,
         0.4453,  0.3530, 10.8300,  2.1303,  0.4190,  0.3322, 10.7820,  1.6812,
         0.4172,  0.3307, 10.7730,  2.1190,  0.3304,  6.7693,  2.1190,  0.3304,
        10.7810,  1.6811,  0.4171,  0.3307,  1.0556,  0.3304, 35.8960,  6.7693,
         1.3316,  0.4168], device='cuda:0')


INFO:root:
          Epoch 0022 | Training loss 10.749467 | Training R2 0.705497 | Validation loss 10.502131 | Validation R2 0.683161
          Best loss 10.502131 | Best epoch 0022
          

          Epoch 0022 | Training loss 10.749467 | Training R2 0.705497 | Validation loss 10.502131 | Validation R2 0.683161
          Best loss 10.502131 | Best epoch 0022
          
100%|##########| 672/672 [04:57<00:00,  2.26it/s]


tensor([13.4830, 50.0187, 23.7006,  5.4426, 37.8579, 16.6934,  4.9872, 29.8659,
        16.8404,  6.1422, 29.4313, 17.8592], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([17.8830, 55.3190, 16.9870,  2.1271, 56.8470, 17.5230,  2.1942, 57.7130,
        17.7900,  2.2277, 58.0120, 17.8830], device='cuda:0')


INFO:root:
          Epoch 0023 | Training loss 9.863072 | Training R2 0.752064 | Validation loss 8.892389 | Validation R2 0.772846
          Best loss 8.892389 | Best epoch 0023
          

          Epoch 0023 | Training loss 9.863072 | Training R2 0.752064 | Validation loss 8.892389 | Validation R2 0.772846
          Best loss 8.892389 | Best epoch 0023
          
100%|##########| 672/672 [05:04<00:00,  2.21it/s]


tensor([ 7.5230, 28.2083,  6.9638, 43.5712, 21.4790,  8.5901, 31.1690, 22.5257,
         9.1989, 23.0525,  9.6945], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 9.4905, 23.3070,  3.3688, 62.3120, 21.5050,  3.5695, 43.4420, 24.6750,
         3.5667, 25.0220,  3.6169], device='cuda:0')


INFO:root:
          Epoch 0024 | Training loss 9.174282 | Training R2 0.785484 | Validation loss 8.411671 | Validation R2 0.796742
          Best loss 8.411671 | Best epoch 0024
          

          Epoch 0024 | Training loss 9.174282 | Training R2 0.785484 | Validation loss 8.411671 | Validation R2 0.796742
          Best loss 8.411671 | Best epoch 0024
          
100%|##########| 672/672 [05:27<00:00,  2.05it/s]


tensor([ 6.6192, 48.7858, 18.1505,  3.0380, 40.1961, 15.6837,  3.2039, 16.1147,
         4.3853, 36.3850, 16.3777,  4.8802], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 6.8146, 52.2980, 17.1410,  3.0146, 53.6910, 17.8200,  3.1340, 18.0930,
         3.1820, 52.6960, 17.4970,  3.0772], device='cuda:0')


INFO:root:
          Epoch 0025 | Training loss 8.685740 | Training R2 0.807722 | Validation loss 8.016184 | Validation R2 0.815405
          Best loss 8.016184 | Best epoch 0025
          

          Epoch 0025 | Training loss 8.685740 | Training R2 0.807722 | Validation loss 8.016184 | Validation R2 0.815405
          Best loss 8.016184 | Best epoch 0025
          
100%|##########| 672/672 [06:36<00:00,  1.69it/s]


tensor([28.0076, 38.4045, 34.8358, 31.7636, 25.6699, 40.2830, 32.0294, 26.8187,
        32.5924, 27.0152, 41.3812, 26.7896, 40.9982, 27.1342, 40.3140, 27.6385,
        40.1446, 28.1381, 40.6030, 28.5903, 41.0018, 28.6758, 39.1366],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([66.9750, 34.8300, 30.4850, 27.1220, 19.3810, 59.4210, 39.4690, 28.2220,
        45.4640, 32.5130, 71.1690, 34.2430, 72.7160, 35.0360, 72.9860, 35.1970,
        73.1260, 35.2700, 73.7120, 35.5420, 73.9550, 35.6660, 66.9750],
       device='cuda:0')


INFO:root:
          Epoch 0026 | Training loss 7.138400 | Training R2 0.870128 | Validation loss 6.485291 | Validation R2 0.879179
          Best loss 6.485291 | Best epoch 0026
          

          Epoch 0026 | Training loss 7.138400 | Training R2 0.870128 | Validation loss 6.485291 | Validation R2 0.879179
          Best loss 6.485291 | Best epoch 0026
          
100%|##########| 672/672 [07:31<00:00,  1.49it/s]


tensor([ 5.0187, 57.1289, 47.2150, 40.5393, 16.1645,  8.8857,  3.5374,  0.2072,
        54.3989, 18.5376,  4.5431,  0.7809, 18.6194,  5.7077,  2.1814, 19.8678,
         6.6993,  3.5592, 20.7630,  9.1452,  4.8885, 21.7630,  9.1384,  5.8849,
        10.2070], device='cuda:0', grad_fn=<IndexBackward0>)
tensor([ 4.2077, 59.4320, 46.9050, 38.1280, 14.2580,  7.9220,  3.6186,  1.1171,
        59.4890, 17.3990,  4.4157,  1.1207, 17.1950,  4.3640,  1.1075, 17.2810,
         4.3856,  1.1130, 16.8630,  5.2057,  1.0861, 16.8560,  4.2779,  1.0857,
         4.2077], device='cuda:0')


INFO:root:
          Epoch 0027 | Training loss 7.282791 | Training R2 0.864820 | Validation loss 6.364136 | Validation R2 0.883651
          Best loss 6.364136 | Best epoch 0027
          

          Epoch 0027 | Training loss 7.282791 | Training R2 0.864820 | Validation loss 6.364136 | Validation R2 0.883651
          Best loss 6.364136 | Best epoch 0027
          
100%|##########| 672/672 [07:42<00:00,  1.45it/s]


tensor([ 8.3549, 41.2138, 34.2881, 29.6694, 25.6575, 15.7572, 47.6375, 30.6324,
        18.5476, 50.3466, 31.9987, 18.9066, 51.0366, 43.9544, 18.6381, 43.7398,
        18.6526, 43.3545, 18.4825, 18.2349, 49.8410, 18.2371, 42.3167, 18.1987],
       device='cuda:0', grad_fn=<IndexBackward0>)
tensor([12.5700, 31.8440, 25.2520, 20.7650, 17.3880, 10.5840, 40.6760, 22.8730,
        13.9690, 43.5540, 24.6480, 15.0630, 44.0460, 35.6960, 15.2700, 35.8460,
        15.3380, 35.8950, 15.3590, 15.2690, 43.9280, 15.2390, 35.5910, 15.2300],
       device='cuda:0')


INFO:root:
          Epoch 0028 | Training loss 6.356060 | Training R2 0.897035 | Validation loss 5.972003 | Validation R2 0.897547
          Best loss 5.972003 | Best epoch 0028
          

          Epoch 0028 | Training loss 6.356060 | Training R2 0.897035 | Validation loss 5.972003 | Validation R2 0.897547
          Best loss 5.972003 | Best epoch 0028
          
100%|##########| 672/672 [07:49<00:00,  1.43it/s]


tensor([ 6.9066, 82.5752, 25.0994,  7.8536,  3.5957, 18.4058,  4.5519, 59.7573,
        22.0278,  7.7739,  5.0958, 68.1590,  5.6257, 21.3132,  9.0211,  6.2920,
        67.1497, 21.4841,  9.6844,  6.7338, 66.4735,  9.6566,  6.8122, 64.8901,
        21.4822,  9.3902,  6.9110, 62.5967, 20.4064,  9.1618,  6.9159, 52.6548,
        20.6670,  9.3857,  6.9895, 60.4102, 23.1409,  9.2480], device='cuda:0',
       grad_fn=<IndexBackward0>)
tensor([ 6.1430, 77.5480, 24.0970,  7.7805,  2.9525, 17.9790,  3.0428, 55.8410,
        21.1480,  6.8284,  3.0454, 63.6220,  2.9330, 20.3360,  6.5662,  2.9284,
        63.5080, 20.3310,  6.5647,  2.9278, 63.5290,  6.5668,  2.9287, 62.7370,
        20.0850,  6.4851,  2.8923, 61.1750, 19.5850,  6.3236,  2.8202, 51.6830,
        19.5730,  6.3199,  2.8186, 59.4280, 22.3600,  6.1430], device='cuda:0')


INFO:root:
          Epoch 0029 | Training loss 6.130345 | Training R2 0.904218 | Validation loss 5.422178 | Validation R2 0.915544
          Best loss 5.422178 | Best epoch 0029
          

          Epoch 0029 | Training loss 6.130345 | Training R2 0.904218 | Validation loss 5.422178 | Validation R2 0.915544
          Best loss 5.422178 | Best epoch 0029
          


1 1


In [16]:
fold, model = 1,1
train, validate, test = datasplitter(data_complete, fold, model)
ptnms = ['5010','5011','5012','5013','5014','5015','5016','5209','5210','5211','5212','5409','5410','5411','5412','5413','5414','5415','5416','5602','5603','5604','5605','5606','5607','5608','5609','5610']
test_interp = pd.DataFrame()
for ptnm in ptnms:
    print(ptnm)
    test_interp = pd.concat([test_interp, generate_interp(int(ptnm))], ignore_index=True)
test_interp.to_csv("test_interp.csv", index=False)
test = pd.read_csv("test.csv")
test.loc[test.CYCL > 5, "AMT"] = 0
test.to_csv("test_nodosing.csv", index=False)
x_train = train[["CYCL", "AMT", "TIME", "TFDS"]]
y_train = train["PK_timeCourse"]
x_val = validate[["CYCL", "AMT", "TIME", "TFDS"]]
y_val = validate["PK_timeCourse"]
x_test = test[["CYCL", "AMT", "TIME", "TFDS"]]
y_test = test["PK_timeCourse"]
train

<ipython-input-13-a1547cc7bd9e>:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.1)
<ipython-input-13-a1547cc7bd9e>:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["PTNM"] = df["PTNM"] + str(0.2)
<ipython-input-13-a1547cc7bd9e>:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide

5010
5011
5012
5013
5014
5015
5016
5209
5210
5211
5212
5409
5410
5411
5412
5413
5414
5415
5416
5602
5603
5604
5605
5606
5607
5608
5609
5610


STUD  DSFQ      PTNM  CYCL    AMT   TIME   TFDS  PK_timeCourse  \
0     1000.0   3.0     00001   1.0  296.6    0.0    0.0        20.3820   
1     1000.0   3.0     00001   1.0    0.0   24.0   24.0        73.1480   
2     1000.0   3.0     00001   1.0    0.0  216.0  216.0        19.7640   
3     1000.0   3.0     00001   1.0  288.0  504.0  504.0         3.2199   
4     1000.0   3.0     00001   2.0    0.0  696.0  192.0        23.2100   
...      ...   ...       ...   ...    ...    ...    ...            ...   
7790  3000.0   3.0  001960.1   1.0    0.0   24.0   24.0        53.9210   
7791  3000.0   3.0  001960.2   1.0  217.0    0.0    0.0         1.7780   
7792  3000.0   3.0  001960.2   1.0    0.0   24.0   24.0        53.9210   
7793  3000.0   3.0  001960.3   1.0  217.0    0.0    0.0         1.7780   
7794  3000.0   3.0  001960.3   1.0    0.0   24.0   24.0        53.9210   

             ID  PK_round1  
0     100000001     20.382  
1     100000001     73.148  
2     100000001     19.764  
3     100000001      0.000  
4     100000001      0.000  
...         ...        ...  
7790  300000196     53.921  
7791  300000196      1.778  
7792  300000196     53.921  
7793  300000196      1.778  
7794  300000196     53.921  

[7795 rows x 10 columns]

In [ ]:
import lightgbm as lgb
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from scipy.stats import pearsonr

# Assuming you have trained the model and made predictions on the validation set


params = {
    'boosting_type': 'gbdt',
    'learning_rate': 0.005,
    'num_leaves': 30,
    'num_trees': 1000,
    'objective': 'regression',
    'metric': 'rmse'
}  

model = lgb.LGBMRegressor(
    boosting_type= 'gbdt',
    learning_rate=0.005,
    num_leaves= 30,
    num_trees=1000,
    objective='regression',
    num_boost_round=10000,
    early_stopping_rounds=50, 
    verbose_eval=100)

model.fit(x_train,y_train,eval_set=[(x_test,y_test),(x_train,y_train)],
          verbose=20,eval_metric='rmse')

# train_data = lgb.Dataset(train_x, train_y)

# # Train the model
# model = lgb.train(train_data,boosting_type= 'gbdt',
#     learning_rate=0.005,
#     num_leaves= 30,
#     num_trees=1000,
#     objective='regression',
#     metric='rmse',num_boost_round=10000, 
#     valid_sets=[train_data, (val_x, val_y)], 
#                   early_stopping_rounds=50, 
#                   verbose_eval=100)

y_pred = model.predict(x_test)

# Evaluate the model using RMSE
rmse = mean_squared_error(y_test, y_pred, squared=False)
print(f"RMSE: {rmse:.4f}")

y_pred = model.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("R2 Score: {:.2f}".format(r2))

pearson_corr, _ = pearsonr(y_test, y_pred)
print(f"Pearson correlation coefficient: {pearson_corr:.4f}")

/usr/local/lib/python3.8/dist-packages/lightgbm/engine.py:118: UserWarning: Found `num_trees` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))
/usr/local/lib/python3.8/dist-packages/lightgbm/engine.py:123: UserWarning: Found `early_stopping_rounds` in params. Will use it instead of argument
  warnings.warn("Found `{}` in params. Will use it instead of argument".format(alias))


Training until validation scores don't improve for 50 rounds.
[20]	training's l2: 335.381	training's rmse: 18.3134	valid_0's l2: 341.92	valid_0's rmse: 18.4911
[40]	training's l2: 288.729	training's rmse: 16.992	valid_0's l2: 313.52	valid_0's rmse: 17.7065
[60]	training's l2: 250.552	training's rmse: 15.8288	valid_0's l2: 293.22	valid_0's rmse: 17.1237
[80]	training's l2: 219.233	training's rmse: 14.8065	valid_0's l2: 279.207	valid_0's rmse: 16.7095
[100]	training's l2: 193.427	training's rmse: 13.9078	valid_0's l2: 270.042	valid_0's rmse: 16.433
[120]	training's l2: 172.24	training's rmse: 13.124	valid_0's l2: 264.441	valid_0's rmse: 16.2616
[140]	training's l2: 154.808	training's rmse: 12.4422	valid_0's l2: 262.174	valid_0's rmse: 16.1918
[160]	training's l2: 140.481	training's rmse: 11.8525	valid_0's l2: 262.269	valid_0's rmse: 16.1947
[180]	training's l2: 128.623	training's rmse: 11.3412	valid_0's l2: 264.735	valid_0's rmse: 16.2707
Early stopping, best iteration is:
[149]	training

##LSTM Model

In [17]:
lstm_train_x = train[["CYCL", "AMT", "TIME", "TFDS", "PK_round1"]]
lstm_train_y = train[["PK_timeCourse"]]
twentyelem = torch.flatten(torch.tensor(lstm_train_x[["TIME", "PK_round1"]].head(20).values, dtype=torch.float32))

In [18]:
class LSTM(nn.Module):
      def __init__(self, input_dim, hidden_dim, output_dim):
        super(LSTM, self).__init__()
        
        self.hidden_dim = hidden_dim
        
        # LSTM layers 1 and 2
        self.lstm1 = nn.LSTM(input_dim, hidden_dim, batch_first=True)
        self.lstm2 = nn.LSTM(hidden_dim, hidden_dim, batch_first=True)
        
        # Decoder
        self.decodernet = nn.Sequential(
            nn.Linear(hidden_dim, 128),
            nn.ReLU(), 
            nn.Linear(128, output_dim))
        init_network_weights(self.decodernet)
      
      def forward(self, x):
        output, _ = self.lstm1(x)
        output = torch.tanh(output)
        output, _ = self.lstm2(output)
        output = torch.tanh(output)

        #concatenate with PK_cycle1
        print(output)
        #output = torch.cat((output, twentyelem))

        output = self.decodernet(output)

        return output


# Define hyperparameters
input_dim = 5
hidden_dim = 128
output_dim = 1
batch_size = 1
learning_rate = 0.00005
weight_decay = 0.1
num_epochs = 30
seq_len = 5

#initializing everything
model = LSTM(input_dim, hidden_dim, output_dim)
optimizer = torch.optim.Adam(model.parameters(), lr=0.00005, weight_decay=0.1)
criterion = nn.MSELoss()


#train loop
for epoch in range(num_epochs):
  optimizer.zero_grad()
  x = torch.tensor(lstm_train_x.values, dtype=torch.float32)
  y = torch.tensor(lstm_train_y.values, dtype=torch.float32)

  output = model(x)
  loss = criterion(output, y)
  loss.backward()
  optimizer.step()
        
  print('Epoch [{}/{}], Batch [{}/{}], Loss: {:.4f}'.format(epoch+1, num_epochs, epoch+1, len(train), loss.item()**(0.5)))

tensor([[ 0.0355,  0.0066, -0.1040,  ..., -0.0288,  0.0118, -0.0357],
        [ 0.0595, -0.0376, -0.0706,  ...,  0.0110,  0.1287,  0.0078],
        [ 0.0361, -0.0480, -0.1347,  ..., -0.0966,  0.0580,  0.0037],
        ...,
        [-0.0015, -0.0148, -0.1271,  ..., -0.0615,  0.1382,  0.0100],
        [-0.0243,  0.0322, -0.1807,  ..., -0.1305,  0.1057,  0.0029],
        [-0.0006, -0.0152, -0.1270,  ..., -0.0610,  0.1386,  0.0098]],
       grad_fn=<TanhBackward0>)
Epoch [1/30], Batch [1/7795], Loss: 27.7238
tensor([[ 3.5834e-02,  6.6643e-03, -1.0430e-01,  ..., -2.8557e-02,
          1.1468e-02, -3.5510e-02],
        [ 5.9606e-02, -3.7527e-02, -7.0859e-02,  ...,  1.1037e-02,
          1.2809e-01,  8.2480e-03],
        [ 3.5928e-02, -4.8620e-02, -1.3540e-01,  ..., -9.5884e-02,
          5.7191e-02,  4.8181e-03],
        ...,
        [-9.9317e-04, -1.5084e-02, -1.2771e-01,  ..., -6.0964e-02,
          1.3744e-01,  1.0547e-02],
        [-2.3692e-02,  3.2045e-02, -1.8126e-01,  ..., -1.2988e-01

In [ ]:
# for fold in np.arange(1,6):
#   for model in np.arange(1,6):
#      train, validate, test = datasplitter(data_complete, fold, model)
#      runtraining(1000, model, fold, 0.00005, 0.1, 30, 1e-4, True)
#      runpredict(1000, model, fold, 0.00005, 0.1, 30, 1e-4, True)

#      print(fold, model)

In [ ]:
complete_data

NameError: ignored

In [ ]:
for fold in np.arange(1,6):
  for model in np.arange(1,6):
     train, validate, test = datasplitter(data_complete, fold, model)
     

     print(fold, model)